File: pkg/lib/tasks/specification/specification.go
package specification

import "github.com/ischenkx/kantoku/pkg/lib/tasks/specification/typing"

type ResourceSet struct {
	Naming map[int]string
	Types  map[int]typing.Type
}

type IO struct {
	Inputs  ResourceSet
	Outputs ResourceSet
}

type Executable struct {
	Type string
	Data map[string]any
}

type Specification struct {
	// ID is a path to specification, it's expected to be like /a/b/c/task.go
	ID         string
	IO         IO
	Executable Executable
	Meta       map[string]any
}

type TypeWithID struct {
	ID   string
	Type typing.Type
}

File: pkg/lib/gateway/web/app/src/entities/utils/specs/specification.ts
export type Type = {
    name: string
    // TODO: use subTypes
    sub_types: Record<string, Type>
}

export type ResourceSet = {
    naming: {index: number, name: string}[]
    types: {index: number, type: Type}[]
}

export type IO = {
    inputs: ResourceSet
    outputs: ResourceSet
}

export type Executable = {
    type: string
    data: Record<string, any>
}

export type Specification = {
    id: string
    executable: Executable
    io: IO
    meta: Record<string, any>
}
